src/main/js/UserMailConfigurationForm.tsx

```diff
@@ -21,65 +21,215 @@
+  };
+
+  useEffect(() => {
+    getContent((config._links.availableTopics as Link).href)
+      .then(topics => {
+        setTopicsLoading(false);
+        setAvailableTopics(topics);
+      })
+      .catch(error => {
+        setTopicsLoading(false);
+        setError(error);
+      });
+  }, [initialConfiguration]);
+
+  if (error) {
+    return <ErrorNotification error={error} />;
   }
 
-  configChangeHandler = (value: string, name: string) => {
-    this.setState(
-      {
-        [name]: value
-      },
-      () =>
-        this.props.onConfigurationChange(
-          {
-            ...this.state
-          },
-          true
-        )
+  const topicsEqual = (t1: Topic) => {
+    return (t2: Topic) => {
+      return t1.category.name === t2.category.name && t1.name === t2.name;
+    };
+  };
+
+  const languageChangedHandler = (value: string) => {
+    const newConfig = { ...config, language: value };
+    setConfig(newConfig);
+    onConfigurationChange(newConfig, true);
+  };
+
+  const topicChangedHandler = (topic: Topic) => {
+    return (value: boolean) => {
+      let newExcludedTopics;
+      if (value) {
+        if (config.excludedTopics) {
+          newExcludedTopics = config.excludedTopics.filter(other => !topicsEqual(topic)(other));
+        }
+      } else {
+        newExcludedTopics = config.excludedTopics ? [...config.excludedTopics] : [];
+        newExcludedTopics.push(topic);
+      }
+      const newConfig = { ...config, excludedTopics: newExcludedTopics };
+      setConfig(newConfig);
+      onConfigurationChange(newConfig, true);
+    };
+  };
+
+  const categoryChangedHandler = (category: string) => {
+    return (value: boolean) => {
+      let newExcludedTopics;
+      if (value) {
+        if (config.excludedTopics) {
+          newExcludedTopics = config.excludedTopics.filter(other => other.category.name !== category);
+        }
+      } else {
+        newExcludedTopics = config.excludedTopics ? [...config.excludedTopics] : [];
+        availableTopics.topics.filter(topic => topic.category.name === category).forEach(topic => newExcludedTopics.push(topic));
+      }
+      const newConfig = { ...config, excludedTopics: newExcludedTopics };
+      setConfig(newConfig);
+      onConfigurationChange(newConfig, true);
+    };
+  };
+
+  type TopicInCategories = {
+    [name: string]: Topic[];
+  };
+
+  const groupTopics = () => {
+    const categories: TopicInCategories = {};
+    new Set(availableTopics.topics.map(topic => topic.category.name)).forEach(
+      categoryName =>
+        (categories[categoryName] = availableTopics.topics.filter(topic => topic.category.name === categoryName))
     );
+    return categories;
   };
 
-  render() {
-    const { readOnly, t } = this.props;
+  const isTopicSelected = (topic: Topic) => {
+    if (!config?.excludedTopics) {
+      return true;
+    }
+    return !config.excludedTopics.find(topicsEqual(topic));
+  };
+
+  const divideTopicsIntoColumns = () => {
```

Comment:
I wonder why you chose the column layout. How will we handle this form if several new categories are added? Wouldn't a collapsable container per category for example be better? You dividing logic seems way to complicated.

Reply:
I don't think that we will ever get more than three or four categories for a normal installation, and most categories will only have one topic. I'm not sure that collapsable containers are the right thing to use.